File: backend/govinda-common/src/main/java/net/voytrex/govinda/common/api/ErrorResponse.java
/*
 * Govinda ERP - API Error Response
 * Copyright 2024 Voytrex
 * SPDX-License-Identifier: Apache-2.0
 */

package net.voytrex.govinda.common.api;

import io.swagger.v3.oas.annotations.media.Schema;
import java.time.Instant;
import java.util.List;

/**
 * Standard error response for API errors.
 */
@Schema(description = "Standard error response structure returned by the API")
public record ErrorResponse(
    @Schema(
        description = "Machine-readable error code",
        example = "ENTITY_NOT_FOUND",
        requiredMode = Schema.RequiredMode.REQUIRED
    )
    String errorCode,
    
    @Schema(
        description = "Human-readable error message (localized)",
        example = "Person with ID 550e8400-e29b-41d4-a716-446655440000 not found",
        requiredMode = Schema.RequiredMode.REQUIRED
    )
    String message,
    
    @Schema(
        description = "Timestamp when the error occurred",
        example = "2024-01-15T10:30:00Z",
        requiredMode = Schema.RequiredMode.REQUIRED
    )
    Instant timestamp,
    
    @Schema(
        description = "API path where the error occurred",
        example = "/api/v1/masterdata/persons/550e8400-e29b-41d4-a716-446655440000",
        requiredMode = Schema.RequiredMode.REQUIRED
    )
    String path,
    
    @Schema(description = "Optional field-level validation errors. Only present for validation errors.")
    List<ErrorDetail> details
) {
    public ErrorResponse(String errorCode, String message, Instant timestamp, String path, List<ErrorDetail> details) {
        this.errorCode = errorCode;
        this.message = message;
        this.timestamp = timestamp != null ? timestamp : Instant.now();
        this.path = path;
        this.details = details;
    }

    public ErrorResponse(String errorCode, String message, String path) {
        this(errorCode, message, Instant.now(), path, null);
    }
}

File: backend/govinda-common/src/main/kotlin/net/voytrex/govinda/common/api/ErrorResponse.kt
/*
 * Govinda ERP - API Error Response
 * Copyright 2024 Voytrex
 * SPDX-License-Identifier: Apache-2.0
 */

package net.voytrex.govinda.common.api

import java.time.Instant

/**
 * Standard error response for API errors.
 */
data class ErrorResponse(
    val errorCode: String,
    val message: String,
    val timestamp: Instant = Instant.now(),
    val path: String? = null,
    val details: List<ErrorDetail>? = null
)

/**
 * Detailed error information for validation errors.
 */
data class ErrorDetail(
    val field: String?,
    val message: String,
    val rejectedValue: Any? = null
)

/**
 * API response wrapper for paginated results.
 */
data class PageResponse<T>(
    val content: List<T>,
    val page: Int,
    val size: Int,
    val totalElements: Long,
    val totalPages: Int,
    val first: Boolean,
    val last: Boolean
)
